src/ray/core_worker/task_interface.h

```diff
@@ -43,45 +42,136 @@ struct ActorCreationOptions {
+        actor_cursor_(actor_cursor),
+        task_counter_(task_counter),
+        num_forks_(num_forks) {}
 
   /// ID of the actor.
   const ray::ActorID &ActorID() const { return actor_id_; };
 
   /// ID of this actor handle.
   const ray::ActorHandleID &ActorHandleID() const { return actor_handle_id_; };
 
- private:
-  /// Cursor of this actor.
+  /// Language of the actor.
+  const ray::WorkerLanguage ActorLanguage() const { return actor_language_; };
+  /// Descriptor of actor definition.
+  /// e.g. class info for Java actor. module and class info for Python actor.
+  const ray::ActorDefinitionDescriptor &ActorDefinitionDescriptor() const {
+    return actor_definition_descriptor_;
+  };
+
+  /// The unique id of the last return of the last task.
+  /// It's used as a dependency for the next task.
   const ObjectID &ActorCursor() const { return actor_cursor_; };
 
+  /// The number of tasks that have been invoked on this actor.
+  const int TaskCounter() const { return task_counter_; };
+
+  /// The number of times that this actor handle has been forked.
+  /// It's used to make sure ids of actor handles are unique.
+  const int NumForks() const { return num_forks_; };
+
+  std::unique_ptr<ActorHandle> Fork() {
+    std::unique_lock<std::mutex> guard(mutex_);
+    auto new_handle = std::unique_ptr<ActorHandle>(new ActorHandle(
+        actor_id_, ComputeNextActorHandleId(actor_handle_id_, ++num_forks_),
+        actor_language_, actor_definition_descriptor_));
+    new_handle->actor_cursor_ = actor_cursor_;
+    new_actor_handles_.push_back(new_handle->actor_handle_id_);
+    return new_handle;
+  }
+
+  void Serialize(std::string *output) {
+    std::unique_lock<std::mutex> guard(mutex_);
+
+    ray::rpc::ActorHandle temp;
+    temp.set_actor_id(actor_id_.Binary());
+    temp.set_actor_handle_id(actor_handle_id_.Binary());
+    temp.set_actor_language((int)actor_language_);
+    for (auto &item : actor_definition_descriptor_) {
+      temp.add_actor_definition_descriptor(item);
+    }
+    temp.set_actor_handle_id(actor_handle_id_.Binary());
+    temp.set_actor_cursor(actor_cursor_.Binary());
+    temp.set_task_counter(task_counter_);
+    temp.set_num_forks(num_forks_);
+    temp.SerializeToString(output);
+  }
+
+  static std::unique_ptr<ActorHandle> Deserialize(const std::string &data) {
+    ray::rpc::ActorHandle temp;
+    temp.ParseFromString(data);
+    ray::ActorDefinitionDescriptor actor_definition_descriptor;
+    for (auto &item : temp.actor_definition_descriptor()) {
+      actor_definition_descriptor.push_back(item);
+    }
+    return std::unique_ptr<ActorHandle>(new ActorHandle(
+        ray::ActorID::FromBinary(temp.actor_id()),
+        ray::ActorHandleID::FromBinary(temp.actor_handle_id()),
+        (WorkerLanguage)temp.actor_language(), actor_definition_descriptor,
+        ray::ObjectID::FromBinary(temp.actor_cursor()), temp.task_counter(),
+        temp.num_forks()));
```

Comment:
Maybe it's easier to just make `ActorHandle` hold a `rpc::ActorHandle`?

Reply:
Done.